deployments/docker/Dockerfile

```diff
@@ -1,17 +1,16 @@
 FROM golang:latest
 
+# Output location for make build
```

Comment:
If that docker image is for local development then this makes sense, but to make it usable we could just 
wget metrics release. This will be more usable with real production use cases.

ping @darahayes @aliok

Reply:
I think perhaps our Dockerfile should simply copy in a pre built binary. This allows us to do 'production' builds and development builds using the one dockerfile.

In the development scenario we would do the following:

* build the binary from our source code. (using `make build` for example)
* Run the Docker build. As part of the Docker build the binary will be copied into the container image.

In the production scenario we can do the same thing:

* `wget` a production binary from github or wherever.
* Run the docker build. Production binary gets copied into the container image.

This gives us some nice advantages:

* One really simple Dockerfile. Doesn't need any Golang tools installed inside. e.g. we can base it off RHEL, Centos, Atomic, whatever we want.
* Super easy to do local dev builds which will be particularly useful for testing local code in OpenShift
* Very clean production builds with the lowest number of dependencies.

Reply:
I think for those use cases it would be nice to publish to a public registry from CI/CD so people can i.e. `docker pull aerogear/metrics:{release}`.
Though we can also have binaries in github releases too, of course.

Reply:
The one thing to note here is that the Makefile might need some slight changes. We might want two build commands.

* `make build` - builds a binary cross-compiled for OSX and Linux, i.e. it can run on Dev machines and in the Docker Container.
* `make build_prod` - builds a binary specifically for Linux. Can be used in the Docker image.

How does that sound @wtrocki @paolobueno ?

Reply:
@darahayes I was already thinking of using [multi stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) here which is similar to s2i.

If its important to have a build that downloads a prebuilt binary we could have `docker build --target production` do that while something like `docker build --target local` also has a slim base image that copies the binary from the first build stage.

Reply:
@paolobueno I agree. I believe the style I propose will allow us to do that. This creates a very clear and simple build and release process.

If we have this type of Dockerfile, as a developer I can do 2 things:

- Build my local code into a binary and create an image.
- Download an official release binary and create an image.

As maintainers we would have a build + release pipeline as follows:

- We cut an official release using the `make release` command which will generate the compiled binaries and the source tar files. It also creates a git tag, e.g. `1.0.0`. A `1.0.0` release in Github is created and the binaries + source is uploaded.

* Now we create the release image by downloading the official `1.0.0` release binary from Github and running the Docker build. We can then push that release to `aerogear/metrics-api:1.0.0`

This gives us a really simple build process that will work in CI/CD. This would also allow us to push master images on every build.

It guarantees that there is always a release in Github and a corresponding image in Dockerhub

Reply:
> make build - builds a binary cross-compiled for OSX and Linux, i.e. it can run on Dev machines and in the Docker Container.

Do you mean 2 binaries? Binary built on OSX won't run in Linux container.

Are there any projects using a workflow like you suggest?

Reply:
@paolobueno My apologies I think we were both writing replies at the same time. I never heard of this multi build stage feature. I like the sound of it.

However, I am wondering what is the value of building the binary in the docker image because our `make release` step already builds the binaries and uploads to Github. We can drastically simplify our build steps here by just accepting a binary. I think we would be adding extra complexity.

To answer @aliok's quesion, I made a mistake, the cross-compilation would result individual binaries for each platform we compile to.